packages/datadog-plugin-couchbase/src/index.js

```diff
@@ -0,0 +1,138 @@
+}
+
+function onRequestFinish (emitter, span) {
+  let rowsListener // eslint-disable-line prefer-const
+
+  const errorListener = (err) => {
+    span.setTag(Tags.ERROR, err)
+    span.finish()
+    emitter.removeListener('rows', rowsListener)
+  }
+
+  rowsListener = () => {
+    span.finish()
+    emitter.removeListener('error', errorListener)
+  }
+
+  emitter.once('rows', rowsListener)
+  emitter.once('error', errorListener)
+}
+
+function createWrapMaybeInvoke (tracer) {
+  return function wrapMaybeInvoke (_maybeInvoke) {
+    return function maybeInvokeWithTrace (fn, args) {
+      const scope = tracer.scope()
+      const callbackIndex = args.length - 1
+      const callback = args[callbackIndex]
+
+      if (callback instanceof Function) {
+        args[callbackIndex] = scope.bind(callback)
+      }
+
+      return scope.bind(_maybeInvoke).call(this, fn, args)
+    }
+  }
+}
+
+function createWrapQuery (tracer) {
+  return function wrapQuery (query) {
+    return function queryWithTrace (q, params, callback) {
+      const scope = tracer.scope()
+
+      if (params instanceof Function) {
+        params = scope.bind(params)
+      } else if (callback) {
+        callback = scope.bind(callback)
+      }
+
+      return scope.bind(query.call(this, q, params, callback))
+    }
+  }
+}
+
+function createWrapN1qlReq (tracer, config) {
+  return function wrapN1qlReq (_n1qlReq) {
+    return function n1qlReqWithTrace (host, q, adhoc, emitter) {
+      const scope = tracer.scope()
+      const n1qlQuery = q.statement
+      const span = startQuerySpan(tracer, config, n1qlQuery)
+
+      addBucketTag(span, this)
+      onRequestFinish(emitter, span)
+
+      return scope.bind(_n1qlReq, span).apply(this, arguments)
+    }
+  }
+}
+
+function addBucketTag (span, bucket) {
+  span.setTag('couchbase.bucket.name', bucket.name || bucket._name)
+}
+
+function wrapRequests (Class, tracer, config) {
+  this.wrap(Class.prototype, '_n1qlReq', createWrapN1qlReq(tracer, config))
+}
+
+function unwrapRequests (Class) {
+  this.unwrap(Class.prototype, '_n1qlReq')
+}
+
+function wrapCouchbase (Class, tracer, config) {
+  this.wrap(Class.prototype, '_maybeInvoke', createWrapMaybeInvoke(tracer, config))
+  this.wrap(Class.prototype, 'query', createWrapQuery(tracer))
+}
+
+function unwrapCouchbase (Class) {
+  this.unwrap(Class.prototype, '_maybeInvoke')
+  this.unwrap(Class.prototype, 'query')
+}
+
+module.exports = [
+  {
+    name: 'couchbase',
+    versions: ['2.4.0'],
+    file: 'lib/bucket.js',
+    patch (Bucket, tracer, config) {
+      tracer.scope().bind(Bucket.prototype)
+
+      wrapCouchbase.call(this, Bucket, tracer, config)
+      wrapRequests.call(this, Bucket, tracer, config)
+    },
+    unpatch (Bucket, tracer) {
+      tracer.scope().unbind(Bucket.prototype)
+
+      unwrapCouchbase.call(this, Bucket)
+      unwrapRequests.call(this, Bucket)
+    }
+  },
+  {
+    name: 'couchbase',
+    versions: ['2.4.0'],
+    file: 'lib/cluster.js',
```

Comment:
Why are we patching the same methods on `Cluster` and `Bucket`? Will this cause double-wrapping?

Reply:
This won't cause double-wrapping since the `query()` and `maybeInvoke()` that we're wrapping in `Cluster` do not call the `query()` and `maybeInvoke()` in `Bucket`.

If `query()` is called on `Cluster`, it calls `Cluster`'s respective `_n1ql()` which calls `Cluster`'s respective `maybeInvoke()` which queues up `Bucket`'s `_n1qlReq()`. In Couchbase v2.4.x - v2.5.x, if you want to execute an Analytics query on the Cluster, `Cluster`'s `query` calls `Cluster`'s `_cbas()` which then calls `Cluster`'s `_cbasReq()` without going through `maybeInvoke()`.

If `query()` is called on `Bucket`, it calls `Bucket`'s respective `_n1ql()` which calls `Bucket`'s respective `maybeInvoke()` which queues up `Bucket`'s `_n1qlReq()`.